ark/utils/data_utils.py

```diff
@@ -5,6 +5,58 @@
 import xarray as xr
 
 
+def relabel_images(fovs, all_data, label_maps):
+    """ Translates cell-ID labeled images according to the clustering assignment.
+
+    Takes a list of fovs, and relabels each image (array) according to the assignment
+    of cell IDs to cluster label.
+
+    Args:
+        fovs (list):
+            List of fovs to relabel.
+        all_data (pandas.DataFrame):
+            data including fovs, cell labels, and cell expression matrix for all markers.
+        label_maps (xr.DataArray):
+            xarray of label maps for multiple fovs
+    Returns:
+        list:
+            The relabeled arrays (representing images).
+    """
+
+    images = []
+    for fov in fovs:
+        df = all_data.loc[all_data['SampleID'] == fov]
+        labels_dict = dict(zip(df.label, df.cluster_labels))
+        images.append(relabel_img_array(label_maps.loc[label_maps['fovs'] == fov].squeeze(),
+                                        labels_dict))
+    return images
+
+
+def relabel_img_array(labeled_image, labels_dict):
+    """Takes a labeled image and translates its labels according to a dictionary.
+
+    Returns the relabeled array (according to the dictionary).
+
+    Args:
+        labeled_image (numpy.ndarray):
+            2D numpy array of labeled cell objects.
+        labels_dict (dict):
+            a mapping between labeled cells and their clusters.
+    Returns:
+        numpy.ndarray:
+            The relabeled array.
+    """
+
+    img = np.copy(labeled_image)
+    unique_cell_ids = np.unique(labeled_image.values)
+    unique_cell_ids = unique_cell_ids[np.nonzero(unique_cell_ids)]
+
+    for cell_id in unique_cell_ids:
+        img[img == cell_id] = labels_dict.get(cell_id, -1)
```

Comment:
The way this is currently implemented won't work as expected due to overlap between cluster labels and cell_id for small cell_ids.

Let's add some test functions for this and the function above. You can talk to @alex-l-kong and @vacuousplanet to see if any of our current test functions produce test data that makes sense here. Otherwise, to start off you can create a labeled array with say 20 different cells, sequentially labeled, and a corresponding dataframe with 3 different cluster types.

Reply:
oh wow can't believe I missed that! I should definitely take the indices from the input image and not from the output